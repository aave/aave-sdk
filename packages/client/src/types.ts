import type { Transaction, TransactionRequest } from '@aave/graphql';
import type { ResultAsync, TxHash } from '@aave/types';
import type { SigningError } from '../dist';
import type { ValidationError } from './errors';

export type OperationHandler<T extends Transaction = Transaction> = (
  result: T,
) => ResultAsync<
  TxHash,
  SigningError | ValidationError<Exclude<T, TransactionRequest>>
>;
