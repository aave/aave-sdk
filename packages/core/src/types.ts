/**
 * The environment configuration type.
 */
export type EnvironmentConfig = {
  name: string;
  backend: string;
  indexingTimeout: number;
  pollingInterval: number;
};

/**
 * A standardized data object.
 *
 * All GQL operations should alias their results to `value` to ensure interoperability
 * with this client interface.
 */
export type StandardData<T> = { value: T };
