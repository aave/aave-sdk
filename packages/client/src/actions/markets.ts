import type { UnexpectedError } from '@aave/core';
import {
  type Market,
  MarketQuery,
  type MarketReservesRequestOrderBy,
  MarketsQuery,
  type MarketUserState,
  OrderDirection,
  UserMarketStateQuery,
  type UserMarketStateRequest,
} from '@aave/graphql';
import type { ChainId, EvmAddress, ResultAsync } from '@aave/types';
import type { AaveClient } from '../AaveClient';

export const defaultMarketReservesRequestOrderBy: MarketReservesRequestOrderBy =
  {
    tokenName: OrderDirection.Asc,
  };

export type MarketsRequest = {
  /**
   * The markets you want to see based on the chain ids.
   */
  chainIds: ChainId[];
  /**
   * The user viewing the market (e.g., the connected wallet).
   *
   * If not provided, user fields will not be included.
   */
  user?: EvmAddress;
  /**
   * The order by clause for the borrow reserves in the market.
   *
   * @defaultValue { tokenName: OrderDirection.Asc }
   */
  borrowsOrderBy?: MarketReservesRequestOrderBy;
  /**
   * The order by clause for the supply reserves in the market.
   *
   * @defaultValue { tokenName: OrderDirection.Asc }
   */
  suppliesOrderBy?: MarketReservesRequestOrderBy;
};

/**
 * Fetches all markets for the specified chain IDs.
 *
 * ```ts
 * const result = await markets(client, {
 *   chainIds: [chainId(1), chainId(8453)],
 * });
 * ```
 *
 * @param client - Aave client.
 * @param request - The markets request parameters.
 * @returns The list of markets.
 */
export function markets(
  client: AaveClient,
  {
    chainIds,
    borrowsOrderBy = defaultMarketReservesRequestOrderBy,
    suppliesOrderBy = defaultMarketReservesRequestOrderBy,
    user,
  }: MarketsRequest,
): ResultAsync<Market[], UnexpectedError> {
  return client.query(MarketsQuery, {
    request: { chainIds, user },
    borrowsOrderBy,
    suppliesOrderBy,
  });
}

export type MarketRequest = {
  /**
   * The pool address for the market.
   */
  address: EvmAddress;

  /**
   * The chain id the market pool address is deployed on.
   */
  chainId: ChainId;
  /**
   * The user viewing the market (e.g., the connected wallet).
   *
   * If not provided, user fields will not be included.
   */
  user?: EvmAddress;
  /**
   * The order by clause for the borrow reserves in the market.
   *
   * @defaultValue { tokenName: OrderDirection.Asc }
   */
  borrowsOrderBy?: MarketReservesRequestOrderBy;
  /**
   * The order by clause for the supply reserves in the market.
   *
   * @defaultValue { tokenName: OrderDirection.Asc }
   */
  suppliesOrderBy?: MarketReservesRequestOrderBy;
};

/**
 * Fetches a specific market by address and chain ID.
 *
 * ```ts
 * const result = await market(client, {
 *   address: evmAddress('0x87870bca3f3fd6335c3f4ce8392d69350b4fa4e2'),
 *   chainId: chainId(1),
 * });
 * ```
 *
 * @param client - Aave client.
 * @param request - The market request parameters.
 * @returns The market data, or null if not found.
 */
export function market(
  client: AaveClient,
  {
    address,
    chainId,
    user,
    borrowsOrderBy = defaultMarketReservesRequestOrderBy,
    suppliesOrderBy = defaultMarketReservesRequestOrderBy,
  }: MarketRequest,
): ResultAsync<Market | null, UnexpectedError> {
  return client.query(MarketQuery, {
    request: { address, chainId, user },
    borrowsOrderBy,
    suppliesOrderBy,
  });
}

/**
 * Fetches user account market data across all reserves.
 *
 * ```ts
 * const result = await userMarketState(client, {
 *   market: evmAddress('0x1234…'),
 *   user: evmAddress('0x5678…'),
 *   chainId: chainId(1)
 * });
 * ```
 *
 * @param client - Aave client.
 * @param request - The user market state request parameters.
 * @returns The user's market state.
 */
export function userMarketState(
  client: AaveClient,
  request: UserMarketStateRequest,
): ResultAsync<MarketUserState, UnexpectedError> {
  return client.query(UserMarketStateQuery, { request });
}
