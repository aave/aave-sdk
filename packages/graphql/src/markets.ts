import { MarketFragment } from './fragments';
import { graphql, type RequestOf } from './graphql';

export const MarketsQuery = graphql(
  `query Markets($request: MarketsRequest!) {
    value: markets(request: $request) {
      ...Market
    }
  }`,
  [MarketFragment],
);
export type MarketsRequest = RequestOf<typeof MarketsQuery>;

export const MarketQuery = graphql(
  `query Market($request: MarketRequest!) {
    value: market(request: $request) {
      ...Market
    }
  }`,
  [MarketFragment],
);
export type MarketRequest = RequestOf<typeof MarketQuery>;
