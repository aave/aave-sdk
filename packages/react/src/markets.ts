import type { MarketRequest } from '@aave/client/actions';
import {
  type Market,
  MarketQuery,
  MarketsQuery,
  type ReservesRequestOrderBy,
} from '@aave/graphql';
import { type ChainId, type EvmAddress, ZERO_ADDRESS } from '@aave/types';
import type {
  ReadResult,
  Suspendable,
  SuspendableResult,
  SuspenseResult,
} from './helpers';
import { useSuspendableQuery } from './helpers';

export type UseAaveMarketArgs = MarketRequest;

/**
 * Fetch a single Aave Market.
 *
 * This signature supports React Suspense:
 *
 * ```tsx
 * const { data } = useAaveMarket({
 *   address: evmAddress('0x8787…'),
 *   chainId: chainId(1),
 *   suspense: true,
 * });
 * ```
 */
export function useAaveMarket(
  args: UseAaveMarketArgs & Suspendable,
): SuspenseResult<Market | null>;

/**
 * Fetch a single Aave Market.
 *
 * ```tsx
 * const { data, loading } = useAaveMarket({
 *   address: evmAddress('0x8787…'),
 *   chainId: chainId(1),
 * });
 * ```
 */
export function useAaveMarket(
  args: UseAaveMarketArgs,
): ReadResult<Market | null>;

export function useAaveMarket({
  suspense = false,
  address,
  chainId,
  userAddress,
  borrowsOrderBy,
  suppliesOrderBy,
}: UseAaveMarketArgs & {
  suspense?: boolean;
}): SuspendableResult<Market | null> {
  return useSuspendableQuery({
    document: MarketQuery,
    variables: {
      request: { address, chainId },
      includeUserFields: !!userAddress,
      userAddress: userAddress ?? ZERO_ADDRESS,
      borrowsOrderBy,
      suppliesOrderBy,
    },
    suspense,
  });
}

export type UseAaveMarketsArgs = {
  chainIds: ChainId[];
  userAddress?: EvmAddress;
  borrowsOrderBy?: ReservesRequestOrderBy;
  suppliesOrderBy?: ReservesRequestOrderBy;
};

/**
 * Fetch all Aave Markets for the specified chains.
 *
 * This signature supports React Suspense:
 *
 * ```tsx
 * const { data } = useAaveMarkets({
 *   chainIds: [chainId(1), chainId(137)],
 *   userAddress: evmAddress('0x742d35cc...'),
 *   suspense: true
 * });
 * ```
 */
export function useAaveMarkets(
  args: UseAaveMarketsArgs & Suspendable,
): SuspenseResult<Market[]>;

/**
 * Fetch all Aave Markets for the specified chains.
 *
 * ```tsx
 * const { data, loading } = useAaveMarkets({
 *   chainIds: [chainId(1), chainId(137)],
 *   userAddress: evmAddress('0x742d35cc...'),
 * });
 * ```
 */
export function useAaveMarkets(args: UseAaveMarketsArgs): ReadResult<Market[]>;

export function useAaveMarkets({
  suspense = false,
  chainIds,
  userAddress,
  borrowsOrderBy,
  suppliesOrderBy,
}: UseAaveMarketsArgs & {
  suspense?: boolean;
}): SuspendableResult<Market[]> {
  return useSuspendableQuery({
    document: MarketsQuery,
    variables: {
      request: { chainIds },
      includeUserFields: !!userAddress,
      userAddress: userAddress ?? ZERO_ADDRESS,
      borrowsOrderBy,
      suppliesOrderBy,
    },
    suspense,
  });
}
