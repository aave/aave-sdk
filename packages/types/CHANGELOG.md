# @aave/types

## 0.1.0

### Minor Changes

- fb26904: **feat:** first release
