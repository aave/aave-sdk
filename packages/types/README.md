# `@aave/types`

Core types for the Aave SDK

---

**It is not intended to be used directly. Its interface will change without notice, use it at your own risk.**

