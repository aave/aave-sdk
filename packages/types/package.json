{
  "name": "@aave/types",
  "version": "0.1.0",
  "description": "Core types for the Aave SDK",
  "repository": {
    "directory": "packages/types",
    "type": "git",
    "url": "git://github.com/aave/aave-sdk.git"
  },
  "type": "module",
  "main": "dist/index.cjs",
  "module": "dist/index.js",
  "types": "dist/index.d.ts",
  "exports": {
    ".": {
      "import": "./dist/index.js",
      "require": "./dist/index.cjs"
    }
  },
  "typesVersions": {
    "*": {
      "import": [
        "./dist/index.d.ts"
      ],
      "require": [
        "./dist/index.d.cts"
      ]
    }
  },
  "files": [
    "dist"
  ],
  "sideEffects": false,
  "scripts": {
    "build": "tsup"
  },
  "dependencies": {
    "neverthrow": "^8.2.0",
    "type-fest": "^4.41.0"
  },
  "devDependencies": {
    "tsup": "^8.5.0",
    "typescript": "^5.9.2"
  },
  "license": "MIT"
}
