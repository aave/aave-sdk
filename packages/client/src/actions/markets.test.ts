import { assertOk, chainId, evmAddress } from '@aave/types';
import { describe, expect, it } from 'vitest';
import {
  client,
  DEFAULT_MARKET_ADDRESS,
  ETHEREUM_FORK_ID,
  wallet,
} from '../test-utils';
import { market, markets, userMarketState } from './markets';

describe('Given the Aave Protocol v3', () => {
  describe('When fetching markets data', () => {
    it('Then it should be possible to fetch markets for a given chain ID', async () => {
      const result = await markets(client, {
        chainIds: [chainId(1)],
      });

      assertOk(result);

      result.value.forEach((market) => {
        expect(market).toMatchSnapshot({
          totalAvailableLiquidity: expect.any(String),
          totalMarketSize: expect.any(String),
          borrowReserves: expect.any(Array),
          supplyReserves: expect.any(Array),
          eModeCategories: expect.any(Array),
        });
      });
    });
  });

  describe('When fetching a market data for a given address', () => {
    it('Then it should be possible to fetch market data for a given market address and chain ID', async () => {
      const result = await market(client, {
        address: DEFAULT_MARKET_ADDRESS,
        chainId: ETHEREUM_FORK_ID,
      });

      assertOk(result);

      expect(result.value).toMatchSnapshot({
        totalAvailableLiquidity: expect.any(String),
        totalMarketSize: expect.any(String),
        borrowReserves: expect.any(Array),
        supplyReserves: expect.any(Array),
        eModeCategories: expect.any(Array),
      });
    });
  });

  describe('When fetching user market state', () => {
    it('Then it should be possible to fetch user market state for a given user, market address and chain ID', async () => {
      const result = await userMarketState(client, {
        market: DEFAULT_MARKET_ADDRESS,
        chainId: ETHEREUM_FORK_ID,
        user: evmAddress(wallet.account!.address),
      });

      assertOk(result);
      expect(result.value).toMatchSnapshot({
        availableBorrowsBase: expect.any(String),
        currentLiquidationThreshold: expect.any(String),
        eModeEnabled: expect.any(Boolean),
        healthFactor: expect.any(String),
        isInIsolationMode: expect.any(Boolean),
        ltv: expect.any(String),
        netAPY: expect.any(String),
        netWorth: expect.any(String),
        totalCollateralBase: expect.any(String),
        totalDebtBase: expect.any(String),
      });
    });
  });
});
