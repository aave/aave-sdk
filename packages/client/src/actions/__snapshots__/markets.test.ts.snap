// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Given the Aave Protocol v3 > When fetching a market data for a given address > Then it should be possible to fetch market data for a given address 1`] = `
{
  "__typename": "Market",
  "address": "0x87870Bca3F3fD6335C3F4ce8392D69350B4fA4E2",
  "borrowReserves": Any<Array>,
  "chain": {
    "__typename": "Chain",
    "chainId": 99999999,
    "explorerUrl": "https://dashboard.tenderly.co/explorer/vnet/27ff3c60-0e2c-4d46-8190-f5170dc7da8c",
    "icon": "https://statics.aave.com/ethereum.svg",
    "name": "Forked Ethereum",
  },
  "eModeCategories": Any<Array>,
  "icon": "https://statics.aave.com/ethereum.svg",
  "name": "AaveV3ForkedEthereum",
  "supplyReserves": Any<Array>,
  "totalAvailableLiquidity": Any<String>,
  "totalMarketSize": Any<String>,
  "userState": null,
}
`;

exports[`Given the Aave Protocol v3 > When fetching markets data > Then it should be possible to fetch markets for a given chain ID 1`] = `
{
  "__typename": "Market",
  "address": "0x0AA97c284e98396202b6A04024F5E2c65026F3c0",
  "borrowReserves": Any<Array>,
  "chain": {
    "__typename": "Chain",
    "chainId": 1,
    "explorerUrl": "https://etherscan.io",
    "icon": "https://statics.aave.com/ethereum.svg",
    "name": "Ethereum",
  },
  "eModeCategories": Any<Array>,
  "icon": "https://statics.aave.com/ethereum.svg",
  "name": "AaveV3EthereumEtherFi",
  "supplyReserves": Any<Array>,
  "totalAvailableLiquidity": Any<String>,
  "totalMarketSize": Any<String>,
  "userState": null,
}
`;

exports[`Given the Aave Protocol v3 > When fetching markets data > Then it should be possible to fetch markets for a given chain ID 2`] = `
{
  "__typename": "Market",
  "address": "0x4e033931ad43597d96D6bcc25c280717730B58B1",
  "borrowReserves": Any<Array>,
  "chain": {
    "__typename": "Chain",
    "chainId": 1,
    "explorerUrl": "https://etherscan.io",
    "icon": "https://statics.aave.com/ethereum.svg",
    "name": "Ethereum",
  },
  "eModeCategories": Any<Array>,
  "icon": "https://statics.aave.com/ethereum.svg",
  "name": "AaveV3EthereumLido",
  "supplyReserves": Any<Array>,
  "totalAvailableLiquidity": Any<String>,
  "totalMarketSize": Any<String>,
  "userState": null,
}
`;

exports[`Given the Aave Protocol v3 > When fetching markets data > Then it should be possible to fetch markets for a given chain ID 3`] = `
{
  "__typename": "Market",
  "address": "0x87870Bca3F3fD6335C3F4ce8392D69350B4fA4E2",
  "borrowReserves": Any<Array>,
  "chain": {
    "__typename": "Chain",
    "chainId": 1,
    "explorerUrl": "https://etherscan.io",
    "icon": "https://statics.aave.com/ethereum.svg",
    "name": "Ethereum",
  },
  "eModeCategories": Any<Array>,
  "icon": "https://statics.aave.com/ethereum.svg",
  "name": "AaveV3Ethereum",
  "supplyReserves": Any<Array>,
  "totalAvailableLiquidity": Any<String>,
  "totalMarketSize": Any<String>,
  "userState": null,
}
`;

exports[`Given the Aave Protocol v3 > When fetching user market state > Then it should be possible to fetch user market state for a given user, market address and chain ID 1`] = `
{
  "__typename": "MarketUserState",
  "availableBorrowsBase": Any<String>,
  "currentLiquidationThreshold": Any<String>,
  "eModeEnabled": false,
  "healthFactor": Any<String>,
  "isInIsolationMode": false,
  "ltv": Any<String>,
  "netAPY": Any<String>,
  "netWorth": Any<String>,
  "totalCollateralBase": Any<String>,
  "totalDebtBase": Any<String>,
}
`;
